kore/src/Kore/Builtin/Int.hs

```diff
@@ -325,3 +331,31 @@ builtinFunctions =
     log2 n
         | n > 0 = Just (smallInteger (integerLog2# n))
         | otherwise = Nothing
+
+evalEq :: Builtin.Function
+evalEq resultSort arguments@[_intLeft, _intRight] =
+    concrete <|> symbolicReflexivity
+  where
+    mkCeilUnlessDefined termLike
+      | TermLike.isDefinedPattern termLike = Condition.topOf resultSort
+      | otherwise =
+        Condition.fromPredicate (makeCeilPredicate resultSort termLike)
+    returnPattern = return . flip Pattern.andCondition conditions
+    conditions = foldMap mkCeilUnlessDefined arguments
+
+    concrete = do
+        _intLeft <- expectBuiltinInt eqKey _intLeft
+        _intRight <- expectBuiltinInt eqKey _intRight
+        _intLeft == _intRight
+            & Bool.asPattern resultSort
+            & returnPattern
+
+    symbolicReflexivity =
+        case (Recursive.project _intLeft, Recursive.project _intRight) of
+            (_ :< VariableF varLeft, _ :< VariableF varRight) ->
+                varLeft == varRight
+                    & Bool.asPattern resultSort
+                    & returnPattern
```

Comment:
This is not right: `varLeft == varRight` _implies_ that `varLeft ==Int varRight`, but not the other way around! We cannot say anything when `varLeft /= varRight`.

Because variables are always defined, there is also no need to include the side conditions here. On the other hand, if we do include the stronger side conditions above, then we do not need to restrict to variables here; we can compare any two symbolic patterns.